maven-surefire-common/src/main/java/org/apache/maven/plugin/surefire/report/TestSetRunListener.java

```diff
@@ -51,13 +50,13 @@
  * @author Kristian Rosenvold
  */
 public class TestSetRunListener
-    implements RunListener, ConsoleOutputReceiver, ConsoleLogger
+    implements TestRunListener, ConsoleLogger
```

Comment:
still we need implements `ConsoleLogger`?
class `TestSetRunListener` is created in one place in `DefaultReporterFactory#createReporter` as  `TestRunListener` ...

Reply:
Ok, I can wrong

On fork side 
`ForkingRunListener` is used as ConsoleLogger
 instance of `ForkingRunListener` is created by `ForkingReporterFactory#createReporter` 

in `ForkedBooter` I see
```
        forkingReporterFactory = createForkingReporterFactory();
        logger = (ConsoleLogger) forkingReporterFactory.createReporter();
```

but when I see `ConsoleLoggerDecorator` - everything can happen in runtime
so don't spend more time on this ... 
maybe after more time it will be more clear for me

Reply:
The data flow `fork` -> `plugin` is all about reporting events.
The data flow `plugin` -> `fork` is about commands.
So this is Command Event based architecture. Similar wordings are in the enterprise world CQRS and Event Sourcing.

The first data flow usually sends objects in the form of:
- new XxxEvent(new ReportEntry()) - typically test success
- new XxxEvent(<string>) - typically std/out from a Thread of a running test
- new XxxEvent(<string>) - typically console error, debug

Therefore there is such a wording `report` because the provider (InPlugin or fork) sends a report back to the plugin, and the provider is under the plugin in the hierarchy because the plugin is a manager of providers.

Now it is visible in the inheritance
`ForkingRunListener` -> `RunListener`, `TestOutputReceiver`, `ConsoleLogger`.
These are 3 types and three interfaces.

I know that you have a problem with the name of class `ForkingRunListener` but it is not mandatory for the developers because the developers and their code operates with abstractions and their code instantiates the object somewhere but the interfaces are use everywhere. The interfaces have different purpose because we do not want `ParallelComputer` to send TEST_SUCCESS of course not, and the PC sends console internal error if any. This is called information hiding. **So there are reports, all these three interfaces are about reports but different group of reports.**

Regarding `ForkingRunListener`, whatsoever do you like `ForkingReportListener` more?

From the history, you should know how the code was before. These methods of reports were everywhere. They did not have one place like it is not. Their implementation was really everywhere, the encoder did not exist in one place and so it was very hard to maintain the code. Naturally this situation happened because we are OSS, and in the OSS you start from simple use cases and you do not have time to think of encapsulation, information hiding, and such things.

Reply:
@slawekjaranowski 
This line of code `logger = (ConsoleLogger) forkingReporterFactory.createReporter();` does not exist in the second commit.

Reply:
Currently, we have to create a new instance of the listener via `createReporter()` because the Provider `surefire-junit47` has own instance in ThreadLocal. After this has finished in [SUREFIRE-1643](https://issues.apache.org/jira/browse/SUREFIRE-1643) and [SUREFIRE-1860](https://issues.apache.org/jira/browse/SUREFIRE-1860), we can refactor `surefire-junit47` where the ThreadLocal would not be needed and the creator of reporter may turn to getter of singleton.

Reply:
@slawekjaranowski 
If you mean to use `ConsoleLoggerDecorator` in order to cast `Object` to `ConsoleLogger`, then I have to say that it is too heavy procedure. Now the second commit does not use casting types because we have a composition of interfaces, so we return the subinterface.